Fixes #27906 - Always prefer modern facts in Facter

ekohl opened the pull request:
Facter started to introduce structured facts in Facter 2. Since Facter 3, the old facts are deprecated. This introduces helper methods to provide the abstraction and always prefers modern facts.

It does make a change that it no longer considers the lsbdistrelease fact. 5d61c18fa7bdee314ac146ba6b35353c16c63b37 already preferred the os.release.full fact which is present with Facter 2.2 or newer. It should also fall back to the operatingsystemrelease fact which is present since Facter 1.5 and a more reliable fact.

To make things easier, it annotates which for each fact which version of Facter introduced it.

ekohl committed:
Fixes #27906 - Always prefer modern facts in Facter

Facter started to introduce structured facts in Facter 2. Since Facter
3, the old facts are deprecated. This introduces helper methods to
provide the abstraction and always prefers modern facts.

It does make a change that it no longer considers the lsbdistrelease
fact. 5d61c18fa7bdee314ac146ba6b35353c16c63b37 already preferred the
os.release.full fact which is present with Facter 2.2 or newer. It
should also fall back to the operatingsystemrelease fact which is
present since Facter 1.5 and a more reliable fact.

To make things easier, it annotates which for each fact which version of
Facter introduced it.

ofedoren commented on app/services/puppet_fact_parser.rb:
```diff
@@ -11,8 +11,8 @@ def operatingsystem
       if os_name[/debian|ubuntu/i] || os.family == 'Debian'
-        if facts.dig(:os, :distro, :codename).presence || facts[:lsbdistcodename]
-          os.release_name = facts.dig(:os, :distro, :codename).presence || facts[:lsbdistcodename]
+        if distro_codename
+          os.release_name = distro_codename
         elsif os.release_name.blank?
           os.release_name = 'unknown'
         end
```
Isn't this equal to
```ruby
os.release_name = distro_codename || 'unknown'
```

ekohl replied:
Not really because it should only be set to `unknown` if it was blank. The distro code name is an optional fact that typically depends on `lsb-release` to be installed (maybe no longer in Facter 4, not sure). You don't want a host where that's not installed to reset a correct value back to `unknown`.

In the future I think the `unknown` part should be dropped, but that's out of scope for this PR.

ekohl replied:
> In the future I think the `unknown` part should be dropped, but that's out of scope for this PR.

For what it's worth, I dropped it in https://github.com/theforeman/foreman/pull/8450, specifically  f667ef7e85e0d5ba5b1be0ee53840d31897379f1. I'd be happy to submit that as a standalone PR with its own issue.

ofedoren commented on app/services/fact_parser.rb:
```diff
@@ -61,10 +80,14 @@ def has_comment?
     Hostgroup.unscoped.where(:title => hostgroup_title).first_or_create if hostgroup_title.present?
   end
 
+  def hostgroup_name
```
Out of curiosity, shouldn't it be `hostgroup_title`. I see and understand  the pattern you want to achieve, but others `.first_or_create` are based on `names`, and the hostgroup's one is based on `title`.

ares commented on app/services/fact_parser.rb:
```diff
@@ -33,18 +33,37 @@ def operatingsystem
   end
 
   def environment
```
this very soon becomes the puppet plugin only and shouldn't be in the abstract parser, puppet environment will no longer be a core resource

before that happens, it can potentially live in the puppet_fact_parser, but again very shortly the codewill need to live in the foreman_puppet plugin, should we have some formal parsers extensions points to avoid including modules or monkeypatching?

ares commented on app/services/fact_parser.rb:
```diff
@@ -33,18 +33,37 @@ def operatingsystem
 
   def domain
+    name = domain_name
+    Domain.unscoped.where(:name => name).first_or_create if name.present?
```
the `unscoped` is scary here, by sending facts, we can create mismatches (domain from org1 while the host is in org2). I know it's not introduced in this PR, just curious if you have any thoughts how to improve this, when we're touching it

of course it applies everywhere

ekohl replied:
Well, it was in the Puppet fact parser before. Looks like [in foreman_ansible](https://github.com/theforeman/foreman_ansible/blob/5eec05f645eae1b1cde9de2179f4cf5208873ae3/app/services/foreman_ansible/fact_parser.rb#L28-L32) it isn't. My idea was that eventually a FactParser should not interact with any DB models and only be an extract of data. By splitting it in the `domain_name` and `domain` I hope we get move in that direction. Perhaps the `domain` should be deprecated as well.

ares replied:
Well the logic will need to live somewhere, perhaps we'll have some FactMapper service class at some point, the ultimate problem is, we don't know the scope for facts when facts come from the proxy.

ekohl replied:
Agreed, but I've decided to move this out of this PR to reduce the scope. Let's move it to https://github.com/theforeman/foreman/pull/8451 so this PR can get in.

ares commented on app/services/puppet_fact_parser.rb:
```diff
@@ -40,38 +40,38 @@ def operatingsystem
-  def model
-    name = facts[:productname] || facts[:model] || facts[:boardproductname]
+  def model_name
+    # TODO: not sure where model comes from, not Facter
```
could it be something in FDI? perhaps that was a custom fact, that should have been prefixed with `foreman_`, we may need to search in git history

ekohl replied:
It was added in 8ac3ea75454ef2bc105c052e76383f3ceaf82232 way back in 2010 as part of https://projects.theforeman.org/issues/289. I don't think FDI existed back then. My guess is that it was either a custom fact or a bug because it meant the fact `hardwaremodel` (which did exist back in the released Facter version at the time, which was 1.5).

lzap replied:
This does not come from FDI. No idea, let's just remove these.

lzap replied:
I am fine to keep it this way, I really don't know. Perhaps a custom fact, yeah. Maybe it is in use.

ares commented on app/services/puppet_fact_parser.rb:
```diff
@@ -201,37 +205,85 @@ def os_name
     end
   end
+
+  def os_release_full
```
should we have os_ and distro_ methods in abstract parser so all plugins parsers implement in a more standard way? I think you mentioned you wanted to do that, definitely not a must for this PR.

ekohl replied:
That is indeed where I think we should go to. In https://github.com/theforeman/foreman/pull/8450 I did some of that work by using the minor releases from facter.

ares reviewed (commented):
Leaving some comments for just reading, happy to test once it's ready.

ekohl commented:
I opened https://github.com/theforeman/foreman/pull/8451 to split off the discussion of model methods or not. I would prefer to merge that PR first but I'm fine either way. Please take a look.

ekohl commented:
This now removes all changes that are #8451. It still lacks tests, but at least it's a bit smaller to review.

ekohl commented:
> I opened #8451 to split off the discussion of model methods or not. I would prefer to merge that PR first but I'm fine either way. Please take a look.

I changed my mind: let's get this in mergable order first.

ekohl commented:
Test failures related, will look into them.

ekohl commented:
Ok, updated the failing tests. I have updated the PR description with the latest commit message and I think it's ready for proper review now. I think dropping lsbdistrelease is safe and shouldn't affect anyone.

lzap commented on app/services/puppet_fact_parser.rb:
```diff
@@ -190,48 +197,127 @@ def use_legacy_facts?
+  rescue StandardError => e
+    logger.error { "Failed to read the hardware ISA: #{e}" }
+    nil
+  end
```
Please move all those methods into the public section, there are already a few and these can be pretty useful down the road.

ekohl replied:
I consider that out of scope for this PR. A concrete implementation shouldn't introduce additional public methods if the abstract base class doesn't. I've talked to @ares about this and @domitea might work on moving more fact parsers into the base repository. That's a great time to do a real refactor and make it consistent across all fact parsers. https://github.com/theforeman/foreman/pull/8450 and https://github.com/theforeman/foreman/pull/8451 are also better fits for that. Most importantly, I want this to make it into Foreman 2.5.

lzap replied:
The reason for this is I want you to write unit tests for all these methods. Unless they are already covered.

lzap reviewed (changes requested):
All those new methods require tests. This is a huge chance to start testing properly.

I suggest you add a JSON into our static_fixtures that is literally generated on the oldest supported facter you aim for (is that 2.2?) and please state in the filename clearly what is it. What we carry in our fixtures is a terrible mixture of various facts from unknown sources and I know that facter version is in facts themselves, but it should be in the filename (thus test helper) so it is obvious what we are testing.

We already have the following:

* test/static_fixtures/facts/example_3.14.16.json
* test/static_fixtures/facts/example_4.0.52.json

Ideally we would run a set of tests in a each loop for each major version we aim to support. This would be a huge stepup for the future proofing our puppet parsing - it has always been a mess.

lzap commented on app/services/puppet_fact_parser.rb:
```diff
@@ -190,48 +197,127 @@ def use_legacy_facts?
+  rescue StandardError => e
+    logger.warning { "Failed to the read distribution id: #{e}" }
+    nil
+  end
```
All those rescue blocks are useless, dig or presence will either return something or nil.

ekohl replied:
This is false. It was the entire conclusion of https://community.theforeman.org/t/puppet-fact-parser-is-failing-with-string-does-not-have-dig-method/23138

The tl;dr of that was that for compatibility with older code Puppet can stringify fact values. That means the `os` structure is stringified and you can't dig into that.

The minimal reproducer:
```
[1] pry(main)> {'os' => 'some string'}.dig('os', 'name')
TypeError: String does not have #dig method
from (pry):1:in `dig'
```

lzap commented:
Well I am not gonna block this but I feel like you are defining here what is the minimum supported facter version, here in this very patch. And I think it is worth making a fixture that clearly communicates that.

ekohl committed:
Refs #27906 - Add FacterDB test

It should be noted that these tests are fairly slow. On my system they
add a minute of runtime to the tests.

ekohl commented:
I did add tests, but they're fairly slow. I still would like to get this into 2.5 so please take a look.

evgeni commented on bundler.d/test.rb:
```diff
@@ -10,6 +10,7 @@
   gem 'show_me_the_cookies', '~> 5.0', :require => false
   gem 'database_cleaner', '~> 1.3', :require => false
   gem 'launchy', '~> 2.4'
+  gem 'facterdb', '~> 1.6'
```
This seems fine for packaging, as we don't care about tests in packaging ;-)

lzap commented on test/unit/puppet_fact_parser_test.rb:
```diff
@@ -438,8 +438,362 @@ def setup
+  # They are structured primairly based on OS rather than Facter version
+  # because FacterDB doesn't contain facts for every combination
+  describe 'Using FacterDB' do
+    subject { get_parser(get_facterdb_facts(facterversion, os_name, os_major)) }
```
The major problem of why these tests take 1 minute is not necessary in the fact that facterdb is slow, but because this subject is called 12 times /each test/ for each operating system /major, minor/, but it yields to the very same result. Let's cache it and I expect the time to drop significantly.

ekohl replied:
Yes, I saw that too. I also wondered if we should use the private method [read_json_file](https://github.com/voxpupuli/facterdb/blob/90a5d5fd10fdc8401db5d83e58fe2069e8a9b587/lib/facterdb.rb#L27-L38).

lzap commented on test/unit/puppet_fact_parser_test.rb:
```diff
@@ -438,8 +438,362 @@ def setup
+    result = FacterDB.get_facts(filter)
+    raise "No facts found for #{os_name} #{os_major} on Facter #{facterversion}" if result.empty?
+    result.first.dup
+  end
```
This method needs to have a LRU cache - when its called with the very same arguments, it should promptly return the result. Now, ideally it should be stored in the test instance as a member attribute, but I do not know if spec/minitest does not create new class for each individual test. Then in that case use a global variable or a class attribute. Please do not cache all possible combinations, that will take too much memory, simply "memorize" the last input and output.

My second concern is FacterDB's `@database` field. It concatenates huge amount of JSON files into a string in-memory. If I was designing this, I would probably move this to the build phase creating one huge JSON file, so it can only be loaded during runtime, preferably memory-mapped. Anyway, back to the problem, it claims 15MB of memory and it never returns it back. This means after this test suite is done, 15MB of memory is lost until the test process dies and OS reclaims it. Please implement a `after(:suite)` block that will set the class attribute to nil so Ruby VM can use that memory again.

You know what - I have an idea: https://github.com/voxpupuli/facterdb/pull/171

ekohl replied:
I agree and implemented a similar cache in https://github.com/voxpupuli/rspec-puppet-facts/commit/43bb17e9f894421daef73e833ca3ea0a506f8d9c but we can't really use that here.

lzap reviewed (dismissed):
It's okay, apologies for a late reply.

My only concern is that we are adding another minute to our 40-50 minutes test suite. We want to do exactly the opposite. I think I found a way how to shave a lot of the time tho - see my patch.

Now, I think we can merge this patch and if facterdb upstream likes my patch, it will speed up automatically after we bump the version. If not, I would like you to do similar thing in our codebase because I do not want another minute to be added. Are you okay with that?

lzap commented:
Oh wait, can you add that `after(:suite)` block to reclaim memory? And rebase please to retest it's been a while.

ezr-ondrej commented:
Ping @ekohl ? This is missing just the last comment and rebase to run the tests, if I'm reading correctly.

lzap commented:
If you can get back to this @ekohl it would be nice to have this prior doing any refactoring changes that are planned by @domitea

ekohl commented:
Rebased & set FacterDB 1.7 as the minimum version so it can safely call `FacterDB.cleanup`.

lzap reviewed (approved):
I am good, great work and improvement of test coverage.

I assume that performance is now better and it should not add a minute or something to our test suite, I am not testing this. Last Jenkins duration seems okay, 53 mins. I will ensure this does not break Discovery and I am merging. Thanks!

lzap commented:
Discovery works fine. I think we are still in the pre-branching period tho, not merging today.

ekohl commented:
We're past branching. Shall we merge this?

ezr-ondrej reviewed (approved):
Thanks @ekohl ! :+1:
